challenger/src/duplex_challenger.rs

```diff
@@ -67,3 +67,126 @@ impl<F: Field, P: ArrayPermutation<F, WIDTH>, const WIDTH: usize> Challenger<F>
             .expect("Output buffer should be non-empty")
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use p3_field::AbstractField;
+    use p3_goldilocks::Goldilocks;
+    use p3_symmetric::permutation::CryptographicPermutation;
+
+    use super::*;
+
+    const WIDTH: usize = 32;
+
+    type TestArray = [F; WIDTH];
+    type F = Goldilocks;
+
+    struct TestPermutation {}
+
+    impl CryptographicPermutation<TestArray> for TestPermutation {
+        fn permute(&self, input: TestArray) -> TestArray {
+            let mut output = input.clone();
+            output.reverse();
+            output.try_into().unwrap()
+        }
+
+        fn permute_mut(&self, input: &mut TestArray) {
+            input.reverse()
+        }
+    }
+
+    impl ArrayPermutation<F, WIDTH> for TestPermutation {}
+
+    #[test]
+    fn it_works_duplexing() {
+        let permutation = TestPermutation {};
+        let mut duplex_challenger = DuplexChallenger::new(permutation);
+
+        // observe elements before reaching WIDTH
+        (0..WIDTH - 1).for_each(|element| {
+            duplex_challenger.observe_element(F::from_canonical_u8(element as u8));
+            assert_eq!(duplex_challenger.input_buffer.len(), element + 1);
+            assert_eq!(
+                duplex_challenger.input_buffer,
+                (0..element + 1)
+                    .map(|i| F::from_canonical_u8(i as u8))
+                    .collect::<Vec<_>>()
+            );
+            assert_eq!(duplex_challenger.output_buffer, vec![]);
+            assert_eq!(duplex_challenger.sponge_state, [F::ZERO; WIDTH]);
+        });
+
+        // Test functionality when we observe WIDTH elements
+        duplex_challenger.observe_element(F::from_canonical_u8(31));
+        assert_eq!(duplex_challenger.input_buffer, vec![]);
+
+        let should_be_output_buffer: [F; WIDTH] = (0..WIDTH as u8)
+            .rev()
+            .map(|i| F::from_canonical_u8(i))
+            .collect::<Vec<_>>()
+            .try_into()
+            .unwrap();
+        assert_eq!(duplex_challenger.output_buffer, should_be_output_buffer);
+        assert_eq!(duplex_challenger.sponge_state, should_be_output_buffer);
+
+        // Test functionality when observe, over more than WIDTH elements
```

Comment:
I think `Test functionality when observing additional elements` would be clearer